### ISSUE TITLE
Performance issues on tables with many indexes

### ISSUE THREAD

**Issue by chinarosesz** [tiagoln/BulkExtensions#15]
I set the timeout to be 10 minutes and it beats the timeout. The table also has 3 million records, do you know if this behavior is expected for such a large table?

**Comment by tiagoln** [comment]
I've checked the code and noticed I left the timeout hardcoded to 60 seconds, I'll change it back to the timeout set on the context and release this fix in the next update. So if you are using version 1.3.5 or the timeout of the bulk operations will be hardcoded to 60 seconds even if you change it in the context.

Also, can you tell me how many columns the table you are trying to insert have, and if you are using any InsertOptions? With this information I'll benchmark the time it takes here. For comparison, I've benchmarked inserting 2 million rows in a table with 5 columns with BulkInsert and it took less than 10 seconds, with InsertOptions.OutputIdentity it took less than 20 seconds.

**Comment by chinarosesz** [comment]
My current table has five columns and has three indexes on the three columns. I think this is probably the reason why it's slow. I wonder if bulkinsert plays well with index.

**Comment by tiagoln** [comment]
Hmm, indexes make it faster to search tables, but longer to write to, so it may be the cause, and you say there is already 3 million rows on the table. I'll try to mirror this scenario here and see the time it takes to perform a BulkInsert operation.

**Comment by chinarosesz** [comment]
For now, I have removed the indexes and keep the lookup in another table. It seems way faster, but I had to delete all the data and start from scratch again since the lookup or the index table doesn't have any status yet.

**Comment by tiagoln** [comment]
I was reading about this and indeed insert operations take a hit on performance when using indexes, and this slowdown is bigger the more indexes you have on the table. 

The simple BulkInsert operation performs the insert directly on the destination table, while the BulkInsert with IndentityOutput option, and the other bulk operations, uses a staging temporary table(which do not have the constraints or indexes of the destination table), and then merges the data to the destination table. 

I suspect the staged approach do not take the same big hit in performance as the direct approach of BulkInsert. I'll do some tests and if I confirm this theory, I'll make the simple BulkInsert use the same staged approach as the other operations, thus avoiding the big slowdown caused by the indexes.

**Comment by tiagoln** [comment]
Today I've changed operations and unified them in a single class called BulkOperation, which performs all kinds of operations using the same flux, all of them uses staging table to insert, then merge according to the operation type.

I've performed some tests today with the following scenario: Table with a total of 5 columns: 1 primary key, 3 with non clustered indexes, 1 simple DateTime. I've pre-filled the table with 8 million rows. Then I bulk inserted 1 million new rows using `context.BulkInsert(collection)` and the operation concluded successfully within 50~60 seconds when using the new behavior. For comparison, running the same scenario on the same computer using the previous behavior of BulkInsert took 7.41 minutes.

I think the time taken is now reasonable, this fix will be included on the next published version.
